### ISSUE TITLE
CloudRunExecuteJobOperator - logs from inside the container are not shown in Airflow 

### ISSUE THREAD

**Issue by daniel-bartley** [apache/airflow#36963]
### Description

Unlike the operators I'm accustomed to, the default logger is not returning the cloud run jobs logs to the airflow task.

### Use case/motivation

Avoids the context switching of navigating to GCP Cloud or Cloud Logging to access logs.

To submit I'd need guidance on the testing in particular. Well and the debugging of course. I don't understand why the logging is not working.

### Are you willing to submit a PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by dirrao** [comment]
@daniel-bartley You can only see the cloud-run job API calls logging to create and run the jobs. Cloud run job logs are not available in airflow as it is not running in the airflow worker.

**Comment by potiuk** [comment]
Yep. What @dirrao wrote. But If you would like to implement it as Cloud Run Operator feature to monitor, poll and stream such logs to airflow logs - it's entirely ok (as long as CloudRun API would alllow it). We are already doing similar things for example in KubernetesPodOperator, where we basically stream the logs from Pod running elsewhere and log them through airflow logging system.

**Comment by daniel-bartley** [comment]
We currently use KubernetesPodOperator for dbt. 
The intuitive expectations was that the logs would be similar and we were surprised they are not.

Where in the code does KubernetesPodOperator  poll and stream logs? If you can point me to [in the code](https://github.com/apache/airflow/blob/main/airflow/providers/cncf/kubernetes/operators/pod.py) and provide commentary, I might be able to adjust the CR operator.

Not having the logs available in Airflow significantly diminishes airflow's fit-for-purpose for the task.

**Comment by potiuk** [comment]
> We currently use KubernetesPodOperator for dbt.
> The intuitive expectations was that the logs would be similar and we were surprised they are not.

Airflow is collectively created by > 2800 people. I am sometimes surprise the opposite, how coherent a small group of maintainers make those people (like you) to do things. And one of the reason is that they - like you notice differences and are willing to contribute back (like you) for a fantastic piece of software they can use absolutely for free and implement and improve things as partially a way to thank all those 2800 and make it even more fit-for-purpose.

> Where in the code does KubernetesPodOperator poll and stream logs? If you can point me to [in the code](https://github.com/apache/airflow/blob/main/airflow/providers/cncf/kubernetes/operators/pod.py) and provide commentary, I might be able to adjust the CR operator.

You will find a few classes and code here:
https://github.com/apache/airflow/blob/390eacb01a710239e69f71d0c52134b7e87d17d4/airflow/providers/cncf/kubernetes/utils/pod_manager.py

Commentary is quite straightforward. Not sure if you can do  it in similar way, becuase Kuberntes POD API to read logs is quite a bit complex and there is a lot of code to provide resilience, avoid duplication and generally retrieve the logs from running POD in an efficient way, taking into account the weaknesses of the API and fact that there could be multiple containers, in the pod, each with its own logs for example.

But if I were to point you to important part of code, what is important is ONE line.

https://github.com/apache/airflow/blob/390eacb01a710239e69f71d0c52134b7e87d17d4/airflow/providers/cncf/kubernetes/utils/pod_manager.py#L457C33-L457C89

When yo know what next log line you should log just ... log it . The key complexity is that the task run locally from airflow side needs to retrieve the logs from Cloud Run while they are coming and log them one-by-one. Do it in a streaming way ideally so that they show up as they are coming and handle it in an efficient way.

> Not having the logs available in Airflow significantly diminishes airflow's fit-for-purpose for the task.

That's why we ask you (or anyone who will grab the `good first issue`) to make it more fit-for-the-purpose. Generally this is how Open source software is created. Someone sees a need to improve it and ... improves it.

**Comment by potiuk** [comment]
BTW,. You are also absolutely welcome to first develop it for your own use if you will. In Airflow developing Custom Operators  is absolutely first-class-citizen and you as a user are absolutely welcome and encouraged even @daniel-bartley to take a code of any operator and extend it and use it as a custom operator if you miss any functionality. So rest assured - even if it will take some time for you to develop it and contribute back up the quality and expectations of maintainers here, you are absolutely free to run it locally (which would be actually very good for testing). 

Many of our users did that and became contributors after adding a missing feature in their own copy of operators/hooks etc. and contributed them back (again as a form of giving back for the free software they got - and one that they can use without any licence and limitations and with huge community that is welcoming and supportive in general (or so I think). You are absolutely welcome to join this fantastic crowd who created Airflow.

**Comment by dbrtly** [comment]
If cloud logging streamed the logs directly into the airflow gcs logs path, would that be a pathway to seeing the logs in airflow?

**Comment by potiuk** [comment]
> If cloud logging streamed the logs directly into the airflow gcs logs path, would that be a pathway to seeing the logs in airflow?

No. GCS is object storage. It's not a streaming solution. It means that you have an OBJECT to pull, not a file to stream. Once you create an object in GCS it's immutable. You can replace it with new object but you cannot append to it. This is absolutely impossible to stream something to object in GCS because the object does not exist until you finished sending the last byte for it - you cannot read from the object unless it exists.

This is a very basic feature of GCS in case you were not aware of.

**Comment by dbrtly** [comment]
I’m glossing over details of specific jargon. 

Cloud run already streams logs to cloud logging.
I can see a directory in the cloud composer bucket for logs. If cloud logging was configured to sink cloud run logs to this bucket, could airflow read them from there?

effectively is another approach to the logging to use gcp to handle the logs in its normal capacity and then copy them to the gcs bucket and have airflow read them from there. Without airflow actually actively handling the logs.  I potentially dont care if there is latency.

**Comment by potiuk** [comment]
Well. if you find a way to create GCS objects (not stream to GCS - that is impossible) after the job is complete, and you follow naming convention of Airflow (which you have to be wary - is configurable and each airflow instance can have it's own configuration) and you would replace the usual logging that Airflow creaes in this place (for example starting task, seeing result of it etc) then likely yes, you could see logs from Cloud Run logs in the UI (but only AFTER Cloud Run and task completes) - streaming is impossible as mentioned - so basically you would see different logs (locally generated) while the task is running and different logs (only logs from Cloud Run) after the task completes.

Of course this is only only under the condition that airlfow is configured to use GCP as remote logging - Airflow can use (in community providers) - with 10 different remote backends or so (and you can have your own custom logging handler). 

So the case where Airflow uses GCS as remote storate backend. All the other cases will not be handled at all And it's higly unlikely we would accept an operator that only provides a feature that works only in a very specific case (and replacing logs by removing those that are locally generated when Task is starting is also likely no go).

**Comment by potiuk** [comment]
So maybe just to explain how logging works when REMOTE logging uses GCS - because you might not be aware of it. GCS is an object storage and streaming to it is impossible as mentioned. So the way Airflow "GCS logging handler" works is:

1) while tasks are running logs are produced locallly on the worker task is running
2) Airflow UI knows the host name of that worker and worker exposes an API where UI can actually stream the logs directly from the worker (and while task is runnig this is what airflow UI does - it does not use GCS whatsoever) 

3) after task is completed., the complete log is uploaded from the worker to the GCS bucket (this is the first time there is an interaction with GCS for the task). This is (again) because streaming to GCS is not possible, you can only upload a complete object to GCS and once you upload it, you can't append to it, you can only replace it with a new complete object.

4) then Airflow UI - knowing that task is completed will attempt to download the log from GCS. Here it actually does use partial retrieval (for efficient reading only parts of the log that  it displays) - this is possible with object storage, but it's not live streaming -it's merely retriving parts of the object that is there, knowing the complete size of it and it's impossible to read parts of the object until it is fully uploaded to GCS.

BTW. This is not a jargon, you need to understand how Object storage works and how airflow works in all those different cases when object storage is used as logging backend.

If we want Cloud Run logs to be available "live", the only good way to make this works "properly" for all different configurations of remote logging is to be able to stream Cloud Run logs to the worker. and let the worker stream it back to the UI.

**Comment by DrYampy** [comment]
Im compesanting for this by creating another task to fetch the logs, Ill drop in a snippet of code in a bit when I get it working.

Im not sure if would make sense to build this into the CloudRunExecuteJobOperator because i like the operator to be atomic - not everyone needs the logs extracted - rather it might make sense to just make another operator?

**Comment by DrYampy** [comment]
ok So i compensated by having a follow on task fetch the logs - however, if the task fails, you never get back the xcom for the job - furthermore, the xcom doesnt actually point to the execution name but rather the job itself (which I can also fix)

I ended up implementing my own CloudRunExecuteJobOperator that waits for the jov to complete then gets the logs for the job using the GCP logging client - feels like its more lightweight since it is one API call vs a constant polling....

Any interest in my contributing it back in a PR @potiuk?

**Comment by potiuk** [comment]
> I ended up implementing my own CloudRunExecuteJobOperator that waits for the jov to complete then gets the logs for the job using the GCP logging client - feels like its more lightweight since it is one API call vs a constant polling....

> Any interest in my contributing it back in a PR @potiuk?

Sure. If it can be done in generic enough way

**Comment by alexnbferreira** [comment]
@DrYampy hey, did you end up using that Operator code or posting it anywhere? Would be curious to take a look. Although log streaming is nice, I also would not mind to have the operator simply fetch the complete log after the task finishes

**Comment by DrYampy** [comment]
@alexnbferreira so one of the walls I ran into while scaling this was there's a GCP logging api quota you will run afoul of.  

So what happened was I had more than 200 cloud run tasks executing in parallel, and after completion, it would hammer the logging API.

You can of course get the quota raised through google, but because the process is non deterministic I don't recommend going this way.

In our case, we're trying to use cloud run as a stopgap to launching k8s and my job becoming a full time k8s babysitter, so we're gonna scotch tape a different solution that mounts an external table on top of the archived GCP logs so that we can have logging.

**Comment by VICIWUOHA** [comment]
@DrYampy , can you please share the code you used for this?
Perhaps, we can take it up from there and try to build on it. Thank you..
Having this issue currently. @dbrtly , did you abandon this for k8s later on ?

Not sure if someone from the google/ composer team is on this thread to assist in knowing what abject to pull for streaming the logs.

**Comment by dbrtly** [comment]
> [@DrYampy](https://github.com/DrYampy) , can you please share the code you used for this? Perhaps, we can take it up from there and try to build on it. Thank you.. Having this issue currently. [@daniel-bartley](https://github.com/daniel-bartley), [@dbrtly](https://github.com/dbrtly) , did you abandon this for k8s later on ?
> 
> Not sure if someone from the google/ composer team is on this thread to assist in knowing what abject to pull for streaming the logs.

I changed teams. Last I saw, we were putting up with the log availability as is.

**Comment by DrYampy** [comment]
I can share the code that retrieves it, but again, you're gonna be limited by the logging api call quota in the project (default is 60/m).

It's trivial to get it raised, but there's an alternative.

In the execution metadata there's the log URI.

My latest internal version exposes that.

I'll make a commit that does this.

**Comment by DrYampy** [comment]
ok I stopped being a lazy #*&$^&*(%

Check this out:  https://github.com/apache/airflow/blob/0adb5ad1cbee03309ffeffa91ec61722723713f1/providers/google/src/airflow/providers/google/cloud/operators/cloud_run.py#L297

This object has an attribute `metadata`:

The output when you print `metadata` looks something like this:

```
name: "projects/{project}/locations/{region}/jobs/{job_name}/executions/{execution_name}"
uid: "{some_uuid}
generation: 1
labels {
  key: "{key}"
  value: "{value}"
}
labels {
  key: "nb-environment"
  value: "dev"
}
create_time {
  seconds: 1739988001
  nanos: 119240000
}
start_time {
  seconds: 1739988007
  nanos: 971451000
}
completion_time {
  seconds: 1739988070
  nanos: 580709000
}
update_time {
  seconds: 1739988001
  nanos: 119240000
}
launch_stage: GA
job: "{job}"
parallelism: 1
task_count: 1
template {
  containers {
    image: "{container}"
    args: "python3"
    args: "src/entrypoints/sync_api/cache_refresh.py"
    env {
      name: "ENVIRONMENT"
      value: "dev"
    }
    resources {
      limits {
        key: "memory"
        value: "2Gi"
      }
      limits {
        key: "cpu"
        value: "1"
      }
    }
  }
  max_retries: 0
  timeout {
    seconds: 600
  }
  service_account: "{service account}"
  execution_environment: EXECUTION_ENVIRONMENT_GEN2
}
conditions {
  type_: "Started"
  state: CONDITION_SUCCEEDED
  last_transition_time {
    seconds: 1739988007
    nanos: 971451000
  }
}
conditions {
  type_: "Completed"
  state: CONDITION_SUCCEEDED
  last_transition_time {
    seconds: 1739988070
    nanos: 580709000
  }
}
conditions {
  type_: "ContainerReady"
  state: CONDITION_SUCCEEDED
  last_transition_time {
    seconds: 1739988002
    nanos: 344637000
  }
}
conditions {
  type_: "ResourcesAvailable"
  state: CONDITION_SUCCEEDED
  last_transition_time {
    seconds: 1739988002
    nanos: 598798000
  }
}
conditions {
  type_: "Retry"
  state: CONDITION_SUCCEEDED
  message: "System will retry after 0:00:00 from lastTransitionTime for attempt 0."
  last_transition_time {
    seconds: 1739988072
    nanos: 489829000
  }
  severity: INFO
  reason: IMMEDIATE_RETRY
}
observed_generation: 1
succeeded_count: 1
log_uri: "{link to logs in log console}"
etag: "{something}"

```

See the `log_uri` key? thatll take you direct to the logs in log console.

I have found exposing this in the airflow logs to be the most efficient, and doesnt run afoul of the 'call the logs client to get the logs if you have more than 60x operation a min' thing

**Comment by DrYampy** [comment]
If im less of a lazy douche ill make a PR 🤪 - but my kids have dominance over my time 🚸

**Comment by ramonvermeulen** [comment]
> ok I stopped being a lazy #_&$^&_(%
> 
> Check this out:
> 
> [airflow/providers/google/src/airflow/providers/google/cloud/operators/cloud_run.py](https://github.com/apache/airflow/blob/0adb5ad1cbee03309ffeffa91ec61722723713f1/providers/google/src/airflow/providers/google/cloud/operators/cloud_run.py#L297)
> 
> Line 297 in [0adb5ad](/apache/airflow/commit/0adb5ad1cbee03309ffeffa91ec61722723713f1)
> 
>  self.operation = hook.execute_job( 
> This object has an attribute `metadata`:
> 
> The output when you print `metadata` looks something like this:
> 
> ```
> name: "projects/{project}/locations/{region}/jobs/{job_name}/executions/{execution_name}"
> uid: "{some_uuid}
> generation: 1
> labels {
>   key: "{key}"
>   value: "{value}"
> }
> labels {
>   key: "nb-environment"
>   value: "dev"
> }
> create_time {
>   seconds: 1739988001
>   nanos: 119240000
> }
> start_time {
>   seconds: 1739988007
>   nanos: 971451000
> }
> completion_time {
>   seconds: 1739988070
>   nanos: 580709000
> }
> update_time {
>   seconds: 1739988001
>   nanos: 119240000
> }
> launch_stage: GA
> job: "{job}"
> parallelism: 1
> task_count: 1
> template {
>   containers {
>     image: "{container}"
>     args: "python3"
>     args: "src/entrypoints/sync_api/cache_refresh.py"
>     env {
>       name: "ENVIRONMENT"
>       value: "dev"
>     }
>     resources {
>       limits {
>         key: "memory"
>         value: "2Gi"
>       }
>       limits {
>         key: "cpu"
>         value: "1"
>       }
>     }
>   }
>   max_retries: 0
>   timeout {
>     seconds: 600
>   }
>   service_account: "{service account}"
>   execution_environment: EXECUTION_ENVIRONMENT_GEN2
> }
> conditions {
>   type_: "Started"
>   state: CONDITION_SUCCEEDED
>   last_transition_time {
>     seconds: 1739988007
>     nanos: 971451000
>   }
> }
> conditions {
>   type_: "Completed"
>   state: CONDITION_SUCCEEDED
>   last_transition_time {
>     seconds: 1739988070
>     nanos: 580709000
>   }
> }
> conditions {
>   type_: "ContainerReady"
>   state: CONDITION_SUCCEEDED
>   last_transition_time {
>     seconds: 1739988002
>     nanos: 344637000
>   }
> }
> conditions {
>   type_: "ResourcesAvailable"
>   state: CONDITION_SUCCEEDED
>   last_transition_time {
>     seconds: 1739988002
>     nanos: 598798000
>   }
> }
> conditions {
>   type_: "Retry"
>   state: CONDITION_SUCCEEDED
>   message: "System will retry after 0:00:00 from lastTransitionTime for attempt 0."
>   last_transition_time {
>     seconds: 1739988072
>     nanos: 489829000
>   }
>   severity: INFO
>   reason: IMMEDIATE_RETRY
> }
> observed_generation: 1
> succeeded_count: 1
> log_uri: "{link to logs in log console}"
> etag: "{something}"
> ```
> 
> See the `log_uri` key? thatll take you direct to the logs in log console.
> 
> I have found exposing this in the airflow logs to be the most efficient, and doesnt run afoul of the 'call the logs client to get the logs if you have more than 60x operation a min' thing

Thanks, I am also interested in this! It is a nice to have in our current project where we use cloud run jobs to run a [**dbt**](https://github.com/dbt-labs/dbt-core) workload every 15 minutes. Ideally we would have the container logs directly available in the airflow logs, however, reading through this issue this is quite complex. I already investigated a bit, and it is indeed quite easy to get a direct link to the logs in the GCP console which is quite nice. 

Maybe adding a `bool` property to the operator `expose_logging_url`, that logs a direct link to the GCP logging could be a solution? What do you think @potiuk, if this is something desirable I am willing to put in some effort to make this happen.

The idea is then that you just click the direct link from the airflow logs, which will open the cloud run job logs.

**Comment by DrYampy** [comment]
IF you want to change the [CloudRunExecuteJobOperator](https://github.com/apache/airflow/blob/0adb5ad1cbee03309ffeffa91ec61722723713f1/providers/google/src/airflow/providers/google/cloud/operators/cloud_run.py#L352) to actually fetch and print the logs - this does so after completion:

```
    def _wait_for_operation(self, operation: 'operation.Operation'):
        try:
            return operation.result(timeout=self.timeout_seconds)
        except Exception:
            error = operation.exception(timeout=self.timeout_seconds)
            raise AirflowException(error)
        finally:
            
            # This was a process to fetch logs from Google Cloud Logging - it hits the GCP project quota for 60 logging calls/min
            # This can be raised with qupta management, however I opted for just a link to the logs in both Datadog and Logging 
            
            from google.cloud import logging_v2
            client = logging_v2.Client(project=self.project_id)
            execution_name = operation.metadata.name.split('/')[-1]
            print(f"Isolated Execution ID: {execution_name}")
            
            # Define log filter for the specific Cloud Run job execution (run)
            filter_str = f"""
                resource.type="cloud_run_job"
                resource.labels.project_id="{self.project_id}"
                resource.labels.location="{self.region}"
                resource.labels.job_name="{self.job_name}"
                labels."run.googleapis.com/execution_name"="{execution_name}"
            """
            print(f"Full Filter String:\n{filter_str}")
                    
            # Execute the query to fetch logs
            print(f"EXECUTION LOGS BELOW!")
            log_entries = []
            entries = client.list_entries(filter_=filter_str)
            for entry in entries:
                if hasattr(entry, "payload"):
                    if isinstance(entry.payload, str):
                        log_entries.append(entry.payload)
                    elif isinstance(entry.payload, OrderedDict):
                        log_entries.append(entry.payload["status"]["message"])
                    else:
                        log_entries.append(
                            "CORNER: Entry Payload is not a String or OrderedDict:"
                        )
                        log_entries.append(entry.payload)
                else:
                    log_entries.append("CORNER: No Payload in Log Entry:")
                    log_entries.append(entry)

            # Print log entries to Airflow task logs
            for log_entry in log_entries:
                print(log_entry)
```
But again, I dont recommend this bevcause if you consider using this operator at scale where you need to fetch more than 60 executions a minute, youre gonna run afoul of the log api quota, and you will need to request a raise on it

**Comment by dbrtly** [comment]
Thanks for all the discussion on this. I confess I don’t understand it all. I typically see cloud composer configured to enable cloud_logging_only. So this was trickier than I anticipated. 

Thanks to @ramonvermeulen  and @DrYampy  for the code.

**Comment by ramonvermeulen** [comment]
Implementation for a `BaseOperatorLink` just got merged, so at least it is now possible to click through from the Airflow UI directly into the cloud run job logs.

As next steps it might be an option to implement a similar solution as AWS Glue has:
https://github.com/apache/airflow/blob/3198aad1bfd7cbf8b3de56c482dc9d96e00d3a69/providers/amazon/src/airflow/providers/amazon/aws/operators/glue.py#L73

Some sample code in:
https://github.com/apache/airflow/issues/36963#issuecomment-2669566660

Important to keep API quota of `60` per minute in mind:
https://cloud.google.com/logging/quotas#api-limits

Ideally I think the implementation should only fetch the whole logs at once when a job has been finished (if this is supported by the underlaying GCP API). This way it is easier to keep the number of request at a minimum, and you would only run into issues when running more than 60 `CloudRunExecuteJobOperator` at parallel in a minute.

**Comment by VICIWUOHA** [comment]
On my end.. We handled this with custom `post_execute` and `pre_execute` callbacks attached to a custom Operator wrapped around the cloudrun operator to run tasks. (These functions pull from the xcom which contains the metadata dict quoted above)
We found a little race issue when tasks are run in parallel and may pull the same `return_value` .

Welldone @ramonvermeulen :100
